### PR TITLE
Dispatcher

### PR DESCRIPTION
**Pull request by ymatsuda**
@raymondng @eishay 

Please review.

Dispatcher computes the routing table. dispatch methods find instances and their shards. Dispatcher supports a cluster with mixed sharding (ex. 4-way sharding and 8-way sharding in the same cluster).

What it actually does for each instance depends on an function given as an argument. 

DistributedSearchRouter wraps Dispatcher.
plan method decides a distribution plan (which instances to call), and dispatch method uses it to actually send messages to remote machines.

So, if search needs to do two or more scatter-gather operations to process a single search request, it can use the same plan to use the same set of instances. This will help us use cache efficiently.

### CODE REVIEW

**Comment by raymondng** [comment]
Not right now, but would be nice if we can do scatter/gather from "thin" client machines going forward. We can probably have a thin layer to do that. 
Scatter/gather has side effects (e.g. GC) and can be problematic on a search instance. 
Error-handling (say dispatch to different "shard-cluster" on retry) might benefit as well

File: kifi-backend/modules/search/test/com/keepit/search/sharding/DispatcherTest.scala [hunk 1]
```diff
@@ -0,0 +1,129 @@
+package com.keepit.search.sharding
+
+import org.specs2.mutable._
+import com.keepit.common.service.{ServiceType, ServiceStatus}
+import com.keepit.common.zookeeper.{RemoteService, Node, ServiceInstance}
+import scala.concurrent.{Await, Future}
+import scala.concurrent.duration._
+
+class DispatcherTest extends Specification {
```
Comment (raymondng) [inline kifi-backend/modules/search/test/com/keepit/search/sharding/DispatcherTest.scala:8]: Nice :+1:

File: kifi-backend/modules/common/app/com/keepit/search/sharding/Dispatcher.scala [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        // now choose an instance
+        val instances = table(candidateShard)
+        if (instances.size > 0) instances(rnd.nextInt(instances.size)) else null
+      }
+    }
+
+    if (candidate != null && !candidate.serviceInstance.reportedSentServiceUnavailable) {
+      body(candidate.serviceInstance)
+    } else {
+      if (attempts < maxAttempts) {
```
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/search/sharding/Dispatcher.scala:105]: Nice!

File: kifi-backend/modules/common/app/com/keepit/search/sharding/Dispatcher.scala [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        // failed to find an instance. there may an unreachable instance. reload shard->instances map and try again
+        reload()
```
Comment (raymondng) [inline kifi-backend/modules/common/app/com/keepit/search/sharding/Dispatcher.scala:107]: Checking: currently if there's an unreachable shard (due to unreachable instances) in a 'shard-cluster' then we will run out of retries and fail even when the shard is available in a different 'shard-cluster', right?

Comment (ymatsuda) [inline kifi-backend/modules/common/app/com/keepit/search/sharding/Dispatcher.scala:107]: yes.

**Comment by raymondng** [comment]
lgtm

**Pull request merged**